### PR TITLE
feat: python/whatsapp_with_vonage

### LINKED ISSUE #207: 🚀 Feature: Add WhatsApp with Vonage for Python
Appwrite Functions allow you to extend and customize your Appwrite server functionality by executing your custom code. 🤩 You can learn more at our official [Appwrite Functions docs](https://appwrite.io/docs/functions).

Your task is to implement the WhatsApp with Vonage function in Python. You can look at the existing [Appwrite Functions Templates](https://github.com/appwrite/templates) in another coding language.

Your function should behave and be implemented similarly to existing WhatsApp with Vonage Node.js template.

### Tasks summary:

- [ ] Fork & clone Appwrite's [templates](https://github.com/appwrite/templates) repository
- [ ] Create a branch `feat-implement-whatsapp-with-vonage-python`
- [ ] Write code for the function in folder `python/whatsapp_with_vonage/`. 
- [ ] Test the function using Appwrite, **provide a video**.
- [ ] Write a `README.md` explaining what the function does, how to use it, and customize it. Here is [an example](https://github.com/appwrite/templates/tree/main/node/whatsapp-with-vonage).
- [ ] Submit a pull request to our [templates](https://github.com/appwrite/templates) repository

If you need any help, contact us on our [Discord server](https://discord.gg/GSeTUeA).

Are you ready to work on this issue? 🤔 Let us know, and we will assign it to you 😊

**Happy Appwriting!**

### PR DESCRIPTION
**Pull request by probablyArth**
## What does this PR do?

Added `python/whatsapp_with_vonage` function template.

## Test Plan

https://github.com/appwrite/templates/assets/61962871/cc021896-871a-414a-a478-bfd801691403

## Related PRs and Issues
closes #207 

### Have you read the [Contributing Guidelines on issues]

yes

### CODE REVIEW

File: python/whatsapp_with_vonage/src/main.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+from requests import post
+
+
+def main(context):
+    throw_if_missing(
+        dict(environ),
+        [
+            "VONAGE_API_KEY",
+            "VONAGE_ACCOUNT_SECRET",
+            "VONAGE_SIGNATURE_SECRET",
```
Comment (loks0n) [inline python/whatsapp_with_vonage/src/main.py:13]: Let's change the env var name here.

File: python/whatsapp_with_vonage/src/main.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        "text": f'you sent: {body["text"]}',
+        "channel": "whatsapp",
+    }
+
+    url = "https://messages-sandbox.nexmo.com/v1/messages"
+    response = post(
+        url,
+        headers=headers,
+        json=data,
+        auth=(environ["VONAGE_API_KEY"], environ["VONAGE_ACCOUNT_SECRET"]),
```
Comment (loks0n) [inline python/whatsapp_with_vonage/src/main.py:55]: Let's change the env var name here.

File: python/whatsapp_with_vonage/src/main.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            get_static_file("index.html"),
+            200,
+            {"content-type": "text/html"},
+        )
+
+    body = context.req.body
+    headers = context.req.headers
+    token = (headers["authorization"] or "").split(" ")[1]
+    try:
+        decode(token, environ["VONAGE_SIGNATURE_SECRET"], ["HS256"])
```
Comment (loks0n) [inline python/whatsapp_with_vonage/src/main.py:28]: Let's change the env var name here.

File: python/whatsapp_with_vonage/src/main.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
+from .utils import get_static_file, throw_if_missing
+from os import environ
```
Comment (loks0n) [inline python/whatsapp_with_vonage/src/main.py:2]: Small nitpick: in other templates we've used just `import os` and `os.environ["ENVAR"]`
Can we match that?

File: python/whatsapp_with_vonage/src/main.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    throw_if_missing(
+        dict(environ),
+        [
+            "VONAGE_API_KEY",
+            "VONAGE_ACCOUNT_SECRET",
+            "VONAGE_SIGNATURE_SECRET",
+            "VONAGE_WHATSAPP_NUMBER",
+        ],
+    )
+    if context.req.method == "GET":
```
Comment (loks0n) [inline python/whatsapp_with_vonage/src/main.py:17]: Can we add an empty line here

**Review by loks0n — CHANGES_REQUESTED** [review]

**Comment by probablyArth** [comment]
Also, please add the hacktoberfest-accpeted label on the pr, currently it says that my PR is not participating in hacktoberfest.

File: python/whatsapp_with_vonage/src/main.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            {"content-type": "text/html"},
+        )
+
+    body = context.req.body
+    headers = context.req.headers
+    token = (headers["authorization"] or "").split(" ")[1]
+    try:
+        decode(token, os.environ["VONAGE_API_SIGNATURE_SECRET"], ["HS256"])
+    except Exception as e:
+        return context.res.json({"ok": False, "error": "Invalid token!"}, 403)
```
Comment (Meldiron) [inline python/whatsapp_with_vonage/src/main.py:31]: I would recommend error message to be `Payload hash mismatch`, just like in Node version.

File: python/whatsapp_with_vonage/src/main.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    try:
+        throw_if_missing(body, ["from", "text"])
+    except ValueError as e:
```
Comment (Meldiron) [inline python/whatsapp_with_vonage/src/main.py:35]: Might be worth excepting everything

File: python/whatsapp_with_vonage/src/main.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    headers = {
+        "Content-Type": "application/json",
+        "Accept": "application/json",
+    }
+
+    data = {
+        "from": os.environ["VONAGE_WHATSAPP_NUMBER"],
+        "to": body["from"],
+        "message_type": "text",
+        "text": f'you sent: {body["text"]}',
```
Comment (Meldiron) [inline python/whatsapp_with_vonage/src/main.py:47]: Suggesting text to be be:

```js
text: `Hi there! You sent me: ${req.body.text}`,
```

In order to be consistent with Node version

File: python/whatsapp_with_vonage/src/main.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    response = post(
+        url,
+        headers=headers,
+        json=data,
+        auth=(os.environ["VONAGE_API_KEY"], os.environ["VONAGE_API_SECRET"]),
+    )
+
+    if response.ok:
+        result = response.json()
+        context.log(result)
```
Comment (Meldiron) [inline python/whatsapp_with_vonage/src/main.py:61]: Let's remove this log

**Review by Meldiron — CHANGES_REQUESTED** [review]
Let's please address my suggestions.

**Review by loks0n — APPROVED** [review]

**Comment by gewenyu99** [comment]
Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship. 

Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.

**Comment by probablyArth** [comment]
probablyarth

**Pull request merged**